File: backend/routes/roles.py
from fastapi import APIRouter, Depends, HTTPException
from repositories.roles import role_repo
from schemas.role import RoleCreate, RoleUpdate, RoleResponse
from utils.response import ApiResponse

router = APIRouter(prefix="/roles", tags=["Roles"])


@router.get("", response_model=ApiResponse)
async def get_roles(project_id: int = None):
    """Get all roles, optionally filtered by project"""
    if project_id:
        roles = role_repo.get_by_project(project_id)
    else:
        roles = role_repo.get_all()
    return ApiResponse.ok(data=roles)


@router.get("/{role_id}", response_model=ApiResponse)
async def get_role(role_id: int):
    """Get a role by ID"""
    role = role_repo.get_by_id(role_id)
    if not role:
        raise HTTPException(status_code=404, detail="Role not found")
    return ApiResponse.ok(data=role)


@router.post("", response_model=ApiResponse)
async def create_role(role_in: RoleCreate):
    """Create a new role"""
    role = role_repo.create(role_in.model_dump())
    return ApiResponse.ok(
        data=role, message="Role created"
    )


@router.put("/{role_id}", response_model=ApiResponse)
async def update_role(
    role_id: int,
    role_in: RoleUpdate,
):
    """Update a role"""
    role = role_repo.get_by_id(role_id)
    if not role:
        raise HTTPException(status_code=404, detail="Role not found")
    updated = role_repo.update(role_id, role_in.model_dump(exclude_unset=True))
    return ApiResponse.ok(
        data=updated, message="Role updated"
    )


@router.delete("/{role_id}", response_model=ApiResponse)
async def delete_role(role_id: int):
    """Delete a role"""
    role = role_repo.get_by_id(role_id)
    if not role:
        raise HTTPException(status_code=404, detail="Role not found")
    role_repo.delete(role_id)
    return ApiResponse.ok(message="Role deleted")

File: backend/src/routes/roles.js
/**
 * Roles Routes
 */
import { BaseRepository } from '../repositories/base.js';

const repo = new BaseRepository('roles.json');

/**
 * Format success response
 */
function successResponse(data = null, message = 'Success') {
  return { success: true, message, data };
}

/**
 * Format error response
 */
function errorResponse(message, error = null) {
  return { success: false, message, error, data: null };
}

/**
 * Register roles routes
 */
async function rolesRoutes(fastify) {
  // Get all roles
  fastify.get('/', async (request, reply) => {
    try {
      const roles = await repo.findAll();
      return successResponse(roles);
    } catch (error) {
      request.log.error(error);
      return errorResponse('Failed to get roles');
    }
  });

  // Get role by ID
  fastify.get('/:id', async (request, reply) => {
    try {
      const roleId = parseInt(request.params.id, 10);
      const role = await repo.findById(roleId);
      if (!role) {
        reply.code(404);
        return errorResponse('Role not found');
      }
      return successResponse(role);
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to get role');
    }
  });

  // Create role
  fastify.post('/', async (request, reply) => {
    try {
      const role = await repo.create(request.body);
      return successResponse(role, 'Role created');
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to create role');
    }
  });

  // Update role
  fastify.put('/:id', async (request, reply) => {
    try {
      const roleId = parseInt(request.params.id, 10);
      const updated = await repo.update(roleId, request.body);
      if (!updated) {
        reply.code(404);
        return errorResponse('Role not found');
      }
      return successResponse(updated, 'Role updated');
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to update role');
    }
  });

  // Delete role
  fastify.delete('/:id', async (request, reply) => {
    try {
      const roleId = parseInt(request.params.id, 10);
      const deleted = await repo.delete(roleId);
      if (!deleted) {
        reply.code(404);
        return errorResponse('Role not found');
      }
      return successResponse(null, 'Role deleted');
    } catch (error) {
      request.log.error(error);
      reply.code(500);
      return errorResponse('Failed to delete role');
    }
  });
}

export default rolesRoutes;
